Script adaptado para uso no postgres

labanca opened the pull request.

labanca committed:
Script adaptado para uso no postgres

labanca committed:
Criação de tabelas e Carga de Dados.

labanca committed:
Script todo adaptado, mas apresentando lentidão para carregar arquivos csv e criar tabelas.

fjuniorr commented:
@labanca você fez push de todos os commits? 

Inicialmente ao executar `make run` tive um erro de "psycopg2.OperationalError: connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory Is the server running locally and accepting connections on that socket?" que foi resolvido adicionando o argumento `host="localhost"` nas chamadas de `psycopg2.connect`.

Depois dessas alterações tive um erro:

```
make run
Não há tabelas na database dadosmg
Não há tabelas em dadosmg
Traceback (most recent call last):
  File "/Users/fjunior/Projects/postgresql-carga-despesa/main.py", line 178, in <module>
    tables_from_csv(file_paths, con)
  File "/Users/fjunior/Projects/postgresql-carga-despesa/main.py", line 66, in tables_from_csv
    df = pd.read_csv(file, delimiter=';', decimal='.')
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/pandas/util/_decorators.py", line 211, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/pandas/io/parsers/readers.py", line 950, in read_csv
    return _read(filepath_or_buffer, kwds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/pandas/io/parsers/readers.py", line 611, in _read
    return parser.read(nrows)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/pandas/io/parsers/readers.py", line 1778, in read
    ) = self._engine.read(  # type: ignore[attr-defined]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/pandas/io/parsers/c_parser_wrapper.py", line 230, in read
    chunks = self._reader.read_low_memory(nrows)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pandas/_libs/parsers.pyx", line 808, in pandas._libs.parsers.TextReader.read_low_memory
  File "pandas/_libs/parsers.pyx", line 866, in pandas._libs.parsers.TextReader._read_rows
  File "pandas/_libs/parsers.pyx", line 852, in pandas._libs.parsers.TextReader._tokenize_rows
  File "pandas/_libs/parsers.pyx", line 1973, in pandas._libs.parsers.raise_parser_error
pandas.errors.ParserError: Error tokenizing data. C error: Expected 1 fields in line 118, saw 2

make: *** [run] Error 1
```

labanca commented:
![image](https://user-images.githubusercontent.com/7211451/236288210-4f4d7b16-a75d-4e36-9730-bc42e3242051.png)

Pelo git status somente a diferença dos commits é somente na versão do psycopg2 no requirements.txt. Estou executando o `make run` novamente para verificar se o erro acontece aqui também. Como a importação do postgres ainda está muito lenta, tem mais de 20 minutos que está executando.

Dando uma olhada na documentação, o "localhost" parece ser o padrão do parametro quando não especificado, mas há uma nota sobre Unix:
`*host*: database host address (defaults to UNIX socket if not provided)`

Você executou no Mac, certo? Pode ter alguma relação com OS.

Outra possibilidade é um erro devido a arquivo csv mal formatado ou talvez até o decimal sendo "." estar gerando algum problema. O erro `"tokenizing data. C error: Expected 1 fields in line 118, saw 2"` costuma ser um caso de parsing errado gerando mais ou menos valores do que colunas da tabela ou dataframe.

Estou apenas relatando as possibilidades, mas já estou averiguando a questão.

labanca committed:
datapackage usado para carga e DDL. Postgres continua lento.

labanca committed:
erro na conversão de dtypes com datetime64[ns]

labanca committed:
erros de datetime[ns] e NaN em colunas integer resolvido

labanca committed:
atualizando makefile com otimizações

fjuniorr committed:
adiciona instrucoes para abrir docker desktop antes de docker compose up

fjuniorr committed:
adiciona argumento host nas chamadas de psycopg2.connect

Para evitar o erro

(venv) postgresql-carga-despesa (development)*$ make run
Traceback (most recent call last):
  File "/Users/fjunior/Projects/postgresql-carga-despesa/main.py", line 178, in <module>
    create_database(DB_NAME)
  File "/Users/fjunior/Projects/postgresql-carga-despesa/main.py", line 56, in create_database
    con = psycopg2.connect("user=postgres password=postgres")
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fjunior/Projects/postgresql-carga-despesa/venv/lib/python3.11/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.OperationalError: connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?

make: *** [run] Error 1

fjuniorr commented:
~No meu pc o ckanapi está salvando o [recurso CKAN com o arquivo datapackage.json](https://dados.mg.gov.br/dataset/despesa/resource/67bfcc88-66e4-4a65-b46e-4fba788c4496) com o nome `datasets/despesa/data/datapackage-.json` (veja o - no final). Aí eu recebo um erro de que:~

```
The data package has an error: cannot retrieve metadata "datasets/despesa/data/datapackage.json" because 
"[Errno 2] No such file or directory: 'datasets/despesa/data/datapackage.json'" 
make: *** [run] Error 1
```

~Eu removi manualmente o `-` mas passei a receber o erro "FileNotFoundError: [Errno 2] No such file or directory: 'datasets/despesa/data/dm_unidade_orc.csv.gz'" porque os arquivos estavam sendo salvos com o título do recurso.~

Os dois erros acima estavam acontecendo porque a versão especificada do ckanapi estava incorreta no arquivo requirements.txt (deveria ser `4.0` e não `4.04.0`) e a versão 4.4 estava sendo instalada.

fjuniorr committed:
utiliza psycopg2-binary para evitar dependências

fjuniorr committed:
corrige especificacao da versao do pacote ckanapi

Precisamos da versão 4.0 mas instalar a versão 4.04.0 instala a versão 4.4

```bash
python -m pip install ckanapi==4.04.0
ckanapi --version
```

fjuniorr commented:
Consegui executar com sucesso na minha máquina. 

```
Total de linhas das tabelas lidas: 62854083
Tempo Total de execução: 8045.827685832977
```

Pra versão de "produção" faz sentido usarmos https://github.com/splor-mg/atividades/issues/7 e otimizar a performance da carga, mas por agora temos um banco de podemos usar pros testes de conexão com as várias ferramentas.

labanca committed:
atualizando readme.md com instruções de execução do código

labanca committed:
Merge branch 'development' of https://github.com/splor-mg/postgresql-carga-despesa into development

hslinhares committed:
adiciona pycache ao gitignore

fjuniorr committed:
Altera titulo readme

labanca committed:
reorganizando readme.md

hslinhares commented on config.py:
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+import os
```
```suggestion
import pathlib
```

hslinhares commented on README.md:
````diff
@@ -21,3 +59,25 @@ Se o cliente `psql` estiver instalado é possível se conectar ao banco de dados
+
+Fazer download dos arquivos e executar o script de carga de dados:
+```python
+make all
````
```suggestion
make tudo
```

hslinhares commented on config.py:
esse arquivo não deveria ter sido comitado
